File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.fs
open System
open System.IO

let start = DateTime.Now.Ticks

let score100Words =
    File.ReadAllLines("Dictionary.txt")
    |> Seq.filter (fun line -> not (line.Contains("-")) && not (line.Contains(" ")) && not (line.Contains("=")))
    |> Seq.filter (fun line ->
        let mutable scoreSum = 0

        for c in line.ToCharArray() do
            scoreSum <- scoreSum + int c - int 'a' + 1

        scoreSum = 100)
    |> Seq.toList

printfn "%s" (score100Words |> String.concat ",")
printfn $"\nwords number:%d{score100Words.Length}"
printfn $"        time:%d{(DateTime.Now.Ticks - start) / TimeSpan.TicksPerMillisecond}ms"

(*
编码体验：
1、不用写花括号和分号，用缩进来表示代码块
2、bool值取反要用 not，而且对圆括号的要求比较严格，如：not (line.Contains("-")
3、Lambda表达式前面需要加 fun，这个在其他语言中是可以省略的，如：fun score -> score.ToString()
4、字符串中内嵌表达式，如：$"\nwords number:%d{score100Words.Length}"
5、管道符 |> 对于熟悉linux的人来说会很亲切，如：score100Words |> String.concat ","
6、没有 main 方法，直接写代码就可以了
*)

File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.dart
import 'dart:io';

void main(List<String> arguments) {
  final start = DateTime.now();
  final score100Words = File('Dictionary.txt')
      .readAsLinesSync()
      .where((line) => !line.contains("-") && !line.contains(" ") && !line.contains("="))
      .where((word) {
          var scoreSum = 0;
          for (var c in word.runes) {
              scoreSum += c - 'a'.codeUnitAt(0) + 1;
          }
          return scoreSum == 100;
      })
      .toList();
  print(score100Words);
  print('\nwords number: ${score100Words.length}');
  print('        time: ${DateTime.now().difference(start).inMilliseconds}ms');
}

/**
 * 编码体验：
 * 1、不用定义同名的Class
 * 2、没有new关键字
 * 3、自动推断变量类型
 * 4、默认读取文件的方法是异步的 readAsLines，同时也提供了同步的 readAsLinesSync
 * 5、字符串中支持变量表达式
 */
